kingpin/actors/rightscale/mci.py

```diff
@@ -0,0 +1,295 @@
+
+    **Examples**
+
+    .. code-block:: json
+
+        { "actor": "rightscale.mci.Create",
+          "desc": "Create an MCI",
+          "options": {
+              "name": "Ubuntu i386 14.04",
+              "description": "this is our test mci",
+              "images": [
+                {
+                    "cloud": "EC2 us-west-2",
+                    "image": "ami-e29774d1",
+                    "instance_type": "m1.small",
+                    "user_data": "cd /bin/bash"
+                },
+                {
+                    "cloud": "EC2 us-west-1",
+                    "image": "ami-b58142f1",
+                    "instance_type": "m1.small",
+                    "user_data": "cd /bin/bash"
+                }
+              ]
+          }
+        }
+
+    """
+
+    all_options = {
+        'name': (str, REQUIRED, 'The name of the MCI to be created.'),
+        'description': (
+            str, '',
+            'The description of the MCI to be created.'),
+        'images': (
+            list, [],
+            'A list of objects that describe our per cloud image settings.'),
+    }
+
+    def __init__(self, *args, **kwargs):
+        """Validate the user-supplied parameters at instantiation time."""
+
+        super(Create, self).__init__(*args, **kwargs)
+
+        allowed_image_options = (
+            'cloud', 'image', 'instance_type', 'user_data')
+        required_image_options = ('cloud', 'image')
+
+        for image in self.option('images'):
+            # Sanity check that no extra options were passed in
+            for key in image.keys():
+                if key not in allowed_image_options:
+                    raise exceptions.InvalidOptions(
+                        'Invalid option (%s) found in Image %s' % (key, image))
+
+            # Make sure that the required options were passed in
+            for required in required_image_options:
+                if required not in image.keys():
+                    raise exceptions.InvalidOptions(
+                        'Missing option "%s" in Image %s' %
+                        (required, image))
+
+    @gen.coroutine
+    def _get_image_def(self, description):
+        """Returns a fully populated set of Multi Cloud Image settings.
+
+        This method takes in a dictionary with as set of parameters (cloud,
+        image, instance_type, user_data) and returns a fully ready-to-use
+        set of RightScale-formatted parameters to create that image. The method
+        handles discovering the RightScale HREFs for the cloud, image and
+        instance_type options.
+
+        Args:
+            description: A dictionary with the keys: cloud, image,
+            instance_type, user_data
+
+        Returns:
+            A RightScale-formatted array of tuples.
+        """
+
+        # Get our cloud object first -- its required so that we can search for
+        # the image/ramdisk/etc hrefs.
+        cloud = yield self._client.find_by_name_and_keys(
```

Comment:
Each time this is invoked it'll get a new cloud name, right? Just thinking if any caching could optimize this.